src/datahike/db.cljc

```diff
@@ -1193,19 +1193,63 @@
       (update-in [:db-after] with-datom datom)
       (update-in [:tx-data] conj datom)))
 
+
+
+(defn validate-datom-upsert [db ^Datom datom]
+  (when (is-attr? db (.-a datom) :db/unique)
+    (when-let [old (first (-datoms db :avet [(.-a datom) (.-v datom)]))]
+      (when-not (= (.-e datom) (.-e old))
+        (raise "Cannot add " datom " because of unique constraint: " old
+          {:error     :transact/unique
+           :attribute (.-a datom)
+           :datom     datom})))))
+
+
+(defn- with-datom-upsert [db ^Datom datom]
+  (validate-datom-upsert db datom)
+  (let [indexing?     (indexing? db (.-a datom))
+        schema?       (ds/schema-attr? (.-a datom))
+        keep-history? (and (-keep-history? db) (not (no-history? db (.-a datom)))
+                        (not= :db/txInstant (.-a datom)))]
+    (cond-> db
+      ;; Optimistic remove of the schema entry
+      schema? (try
```

Comment:
What is the benefit of this optimistic update?

Reply:
In the end the goal is to remove the schema entry old value if there is one. But we don't know if there is an old version or not, so we try to remove it anyway and if it fails then we just go on.

In the old code we had the following to check whether there is an old version of the schema entry or not.
`(if-some [removing ^Datom (first (-search db [(.-e datom) (.-a datom) (.-v datom)]))]`
But this search is expensive and it is what we wanted to avoid in the first place.

Reply:
Ok, that makes sense. Thanks.